chore/refactor(e2e tests): Solidify Contract for and Cleanup WaitForPodsRunningReady

mkarrmann opened the pull request:
#### What type of PR is this?
/kind cleanup
/kind flake

#### What this PR does / why we need it:

This cleans up the the test helper function `WaitForPodsRunningReady`. In particular, concretely defines the expected behavior for this function, rewrite to follow this expected behavior, and remove the log message that caused the original confusion.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #111092

#### Special notes for your reviewer:

This is an extremely simple issue that's remained open for a quite a while. With this being my first K8s PR, the simplicity was appealing to me! PRs have been opened to address this issue, but closed because agreement wasn't reached on how exactly to approach it. Of course, to agree on what to log, there needs to be clarity on what this function is doing. Unfortunately, I don't think the contract for this function has ever been well-defined.

Firstly, what exactly does it mean to have one parameter specify the minimum of Ready Pods, another specify the maximum number of Pods that are allowed to not be Ready, but also document that this function waits for all Pods are either Ready or Failed? I have no idea how that's supposed to work before reading the code! However, reading the code doesn't clarify much.

The clearest way to see the problem with the previous code is to consider: "What happens if every Pod in this Namespace fails?" The code will remain polling for the duration of timeout. Once it exits, then `notReady` will be 0, and this function will return without an error! *Surely* this can't be the intended behavior. This code has been refactored over the past few years, but this same problem has existed since this file was created 5 years ago. This is just one example of why this code is problematic- simply incrementing `notReady` for failed Pods wouldn't completely fix things in my mind.

Ultimately, I needed to decide what the contract for this function should be. I chose one that balanced similarity to previous behavior with what I felt is expected behavior:

```

WaitForPodsRunningReady waits up to timeout for the following conditions:
 1. At least minPods Pods in Namespace ns are Running and Ready
 2. No more than allowedNotReadyPods Pods in Namespace ns are not either
    Ready, Succeeded, or Failed with a Controller.

```

Once I established this, it became clear that a bunch of the logic was entirely unecessary. In fact, I felt like that specific log line from the original Issue lost its purpose now that the presence of non-Ready Pods isn't treated as exceptional behavior, and I just removed it.

One notable consequence of establishing this contract is that this function now exits earlier than it used to in certain situations (roughly speaking, previously a single non-Ready Pod in the Namespace would make the function continue polling until `timeout`). Considering how amgiguous the contract for this function was previously, there's certainly risk that some tests in some environments implicitely rely upon this function taking longer than it needs to. All pertitent Linux e2e tests seem to be passing when run against a local WSL cluster, and I suppose we'll see about the CI tests. I assume that this sort of risk is considered acceptable?

#### Update 4/14

Based upon feedback, I split `WaitForPodsRunningReady` into two functions:

1. `WaitForPodsRunningReady` is now substantially simplied. In addition to the above simplifications, I removed `allowedNotReadyPods` entirely, since it was unused by almost all of the callers.
2. A new function, `WaitForAlmostAllPodsReady`, which is equivalent in behavior to the previous `WaitForPodsRunningReady`. I just kept some light cleanup I did.

Notably, I opted to copy-and-paste code between the functions instead of trying to make it DRY. Since we're scared to touch `WaitForAlmostAllPodsReady` but have more confidence in what the semantics of `WaitForPodsRunningReady` are, it seems best to not couple these two functions at all.

#### Does this PR introduce a user-facing change?
No
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
N/A

mkarrmann committed:
Cleanup WaitForPodsRunningReady: fail for bad pods and reword log message

mkarrmann committed:
Improve logging and comments in WaitForPodsRunningReady

mkarrmann committed:
Explicitity set contract for WaitForPodsRunningReady, and rewrite accordingly

mkarrmann commented on test/e2e/framework/pod/wait.go:
```diff
@@ -153,41 +155,31 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
 				succeededPods = append(succeededPods, pod.Name)
 			case pod.Status.Phase == v1.PodFailed:
 				// ignore failed pods that are controlled by some controller
+				// TODO either document why failures with controllers are allowed while
```
I have absolutely no idea why there exists this special carve-out for Failed Pods with vs without a Controller. Does anyone know any reason to not treat all Failed Pods as failures? This seemed like the sort of thing that was added for a specific reason, and felt out of scope to determine whether or not was appropriate to change.

pohly replied:
Let's not add TODOs without a clear path towards resolving them.

I suspect the motivation might have been that failed pods owned by a controller will get deleted and recreated, but I am not sure.

mkarrmann replied:
Will remove! I've always had a strong bias towards actively calling out "no one who has recently looked at this code is completely sure why this is written this way", since it makes immediately clearer to readers the distiction between "core logic of the function" and "logic which hasn't been changed only because people are scared to touch it". It would have saved me trouble to have this explicitily called out when reading this code!

Of course, I'm more than happy to defer to you for such style concerns.

pohly replied:
I don't think we as maintainers later on add "what the heck is this code doing" comments to existing code just because we touch it and don't understand what the original author meant with it. Unless something is very obviously broken, the code defines the behavior of a method and shouldn't be changed without considering all implications.

It is not unlikely that you will run into situations like that. Happens to me all the time... :smile: But then try to clarify with the original author or people who might know via some other channels (Slack or in a review comment on your own PR) instead of adding TODO comments.

mkarrmann replied:
Haha I appreciate that response, thank you!

mkarrmann commented on test/e2e/framework/pod/wait.go:
```diff
@@ -196,12 +188,11 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
 			var buffer strings.Builder
 			buffer.WriteString(fmt.Sprintf("Expected all pods (need at least %d) in namespace %q to be running and ready (except for %d).\n", minPods, ns, allowedNotReadyPods))
```
Looking through Git history, it seems like this file used to have quite a bit of logging, but not so much anymore. I considered adding additional logging here for failed Pods, but decided that only have this returned error string is most consistent with the style I see elsewhere.

mkarrmann commented on test/e2e/framework/pod/wait.go:
```diff
@@ -153,41 +155,31 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
 			Pods:                   podList.Items,
 		}, nil
 	})).WithTimeout(timeout).Should(framework.MakeMatcher(func(s *state) (func() string, error) {
-		replicas, replicaOk := int32(0), int32(0)
```
One of the more important changes is that I'm no longer bothering to inspect the status of the Replicas. The only thing this was really doing was potentially prevent the function from returning.

pohly replied:
I suspect that was present for a reason. Not checking this anymore might break the use case of the original author.

mkarrmann commented on test/e2e/framework/pod/wait.go:
```diff
@@ -153,41 +155,31 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
+				// 	failures without controllers are not, or remove this check
 				if metav1.GetControllerOf(&pod) == nil {
-					failedPods = append(failedPods, pod)
+					badPods = append(badPods, pod)
```
This variable used to be called `badPods`, but then was changed to `failedPods`. If it doesn't actually include all `failedPods` and just the one's we've decided are especially bad, then `badPods` seems like a better name.

aojea commented:
/ok-to-test

We should also do the exercise of understanding if the existing consumers of this function are not making some assumption about current semantics, everytime we change one of these functions it has a considerable impact in the form of subtle flakes that take also releases to detect

```
grep -nr WaitForPodsRunningReady test/
test/e2e/cloud/gcp/node_lease.go:100:                   err := e2epod.WaitForPodsRunningReady(ctx, c, metav1.NamespaceSystem, systemPodsNo, 0, framework.PodReadyBeforeTimeout)
test/e2e/cloud/gcp/resize_nodes.go:102:                         err := e2epod.WaitForPodsRunningReady(ctx, c, metav1.NamespaceSystem, systemPodsNo, 0, framework.PodReadyBeforeTimeout)
test/e2e/e2e.go:229:    if err := e2epod.WaitForPodsRunningReady(ctx, c, metav1.NamespaceSystem, int32(framework.TestContext.MinStartupPods), int32(framework.TestContext.AllowedNotReadyNodes), timeouts.SystemPodsStartup); err != nil {
test/e2e/windows/host_process.go:660:           e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, timeout)
test/e2e/windows/host_process.go:714:           e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, 3*time.Minute)
test/e2e/windows/kubelet_stats.go:63:                           err = e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 10, 0, timeout)
test/e2e/windows/kubelet_stats.go:155:                          err = e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 3, 0, timeout)
test/e2e/windows/hyperv.go:98:          e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, timeout)
test/e2e/framework/pod/wait.go:102:// WaitForPodsRunningReady waits up to timeout to ensure that all pods in
test/e2e/framework/pod/wait.go:112:func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns string, minPods, allowedNotReadyPods int32, timeout time.Duration) error {
test/e2e/scheduling/priorities.go:112:          err = e2epod.WaitForPodsRunningReady(ctx, cs, metav1.NamespaceSystem, int32(systemPodsNo), 0, framework.PodReadyBeforeTimeout)
test/e2e/common/node/container_probe.go:615:            err := e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, f.Timeouts.PodStart)
test/e2e/common/node/container_probe.go:697:            err := e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, f.Timeouts.PodStart)
test/e2e/common/node/container_probe.go:1362:           err := e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, f.Timeouts.PodStart)
test/e2e/common/node/container_probe.go:1455:           err := e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, f.Timeouts.PodStart)
test/e2e/common/node/pods.go:876:               err := e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 3, 0, f.Timeouts.PodStart)
```

Thanks for doing this

aojea commented:
/assign @aojea @BenTheElder @pohly 
/sig testing

mkarrmann commented:
@aojea Agreed! :) Thankfully, all but one calls to this functions follow the pattern of:

1. Create exactly `n` Pods
2. Call this function with `minPods=n` and `allowedNotReadyPods=0`
3. Seem to be expecting that exactly `n` Pods are Ready after this function is called.

That is, my change should make this function more consistent with the semantics of these callers. Of course, it's difficult to eyeball the code and completely confirm that none are subtely assuming that the full `timeout` is waited in specific failure scenarios, but that'd be more of a problem with the tests themselves.

The only call that I'd be a bit concern about is the one which is used when setting up the suite in `e2e.go`, where this function is called for the `kube-system` Namepace:

https://github.com/kubernetes/kubernetes/blob/f8930f980d2986f9e486b04c14c3e93e57bdbe12/test/e2e/e2e.go#L225-L237

The only reason that it's concerning is that the values of `minPods` and `allowedNotReadyPods` are configurable, and this is the only place where `allowedNotReadyPods` may take a non-0 value. It's conceivable that someone has configured these flags in their test run poorly, and their tests aren't flaky only because this function simply blocks for the entire duration of `timeout` and doesn't always return an error when it should. But again, that'd be a problem with how the test is configured.

The following call to `waitForDaemonSets` should at least help to mitigate this concern.

mkarrmann commented:
@pohly @BenTheElder @aojea Anything I can/should do to push this along (no rush on my end, of course)?

pohly reviewed (commented):
I agree that this is a complex function with insufficient documentation. I'm less sure how to fix this.

Perhaps create a variant of this function for the common case that [you identified](https://github.com/kubernetes/kubernetes/pull/124205#issuecomment-2041122623), simplify the implementation of that function, and then use the simpler variant in those tests where it is clear that it is the right fit?

mkarrmann commented:
@aojea @pohly Understood! I had a lot of back-and-forth on how much of a refactor to actually do, since I originally intended to merely fix the Issue about the logging. I can definitely appreciate that this amount of changes just doesn't seem worth the risk of dealing with flaky tests.

As suggested, I'll handle the single case which is more concerning that the others separetly. That will actually allow for even greater cleanup.

mkarrmann committed:
Add WaitForAlmostAllPodsReady function, similar to previous WaitForPodsRunningReady function

mkarrmann committed:
Use new WaitForAlmostAllPodsReady function ins e2e setup

mkarrmann committed:
Refactor and simplify WaitForPodsRunningReady; update callers to use new interface

mkarrmann commented:
/retest

mkarrmann committed:
Check err in Windows e2e tests

mkarrmann commented:
/retest

mkarrmann commented on test/e2e/framework/pod/wait.go:
```diff
@@ -211,13 +218,83 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
 	// An error might not be fatal.
-	if err != nil && notReady >= 0 && notReady <= allowedNotReadyPods {
-		framework.Logf("Number of not-ready pods (%d) is below the allowed threshold (%d).", notReady, allowedNotReadyPods)
+	if nOk+len(otherPods) >= minPods && len(otherPods) <= allowedNotReadyPods {
```
This is the only actual logic change I made to this function. I know that the team is hesitant to make any changes to these functions, and I'm happy to undo this change if we want.

However, I felt like this was the one change that had to be kept. Previously, this check would pass and we would return `nil` in obvious failure scenarios such as every Pod failed. Surely that's a logic bug that no test is relying upon. The entire purpose of this function is to check that "enough" Pods are Running and Ready, so hopefully is acceptable to ensure that this function actually checks that.

pohly replied:
I have not had the time to verify the logic of behind this change. Usually that would be a reason to not merge it. I don't doubt that you are doing the right thing, but usually it's the well-intended-but-not-fully-understood (by author and reviewer!) changes which cause regressions.

Do we really need to fix something here?

mkarrmann replied:
Is there a need for fuctions in tests to do what they're supposed to? Perhaps not "need", but it does seem like a good idea! (Apologies if that sounds snarky via text, I really don't mean for it to =D). I think my largest concern I have about the current code is:

1. Many of the Pods this function is waiting on fail
2. This function is bugged so it doesn't return an error
3. Because the Pods failed, another part of the test fails
4. It's difficult to debug why the test failed

That's just hypothetical though.

This is my first K8s PR and I'm just trying to dip my toes in the water. I may have opinions I want to fight for down the road, but this definitely isn't one of them. I'm more than happy to defer to your judgement here.

pohly replied:
> Is there a need for functions in tests to do what they're supposed to?

Sure, but who decides what they are supposed to do? This particular example has a deceptively simple name that implies a certain purpose, but then clearly does a whole lot more than just that.

> Apologies if that sounds snarky via text

No problem :smile:

> I'm more than happy to defer to your judgement here.

I prefer to not change it. It's a simple risk assessment: no-one is complaining about this *in the context of checking the cluster readiness* (which is what this function will be used for going forward). But if we change it, we might break someone and then get complaints.

pohly replied:
The reimplementation of WaitForPodsRunningReady can be simpler and needs to be different - that's okay.

mkarrmann commented on test/e2e/framework/pod/wait.go:
```diff
@@ -99,17 +99,28 @@ func BeInPhase(phase v1.PodPhase) types.GomegaMatcher {
 // If minPods or allowedNotReadyPods are -1, this method returns immediately
 // without waiting.
-func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns string, minPods, allowedNotReadyPods int32, timeout time.Duration) error {
+func WaitForAlmostAllPodsReady(ctx context.Context, c clientset.Interface, ns string, minPods, allowedNotReadyPods int, timeout time.Duration) error {
```
This is almost identical to the previous `WaitForPodsRunningReady`. Most line changes are just cleanup/refactoring. The only actual change is the check at the end for whether to return an error, which I explain there.

This function is only called in the setup function in `e2e.go`, since that's the only place that it isn't clearly safe to change to cleaner function below.

pohly replied:
I think that is where the function originally came from, dating back to 2016 and earlier. That E2E tests then used it just "happened".

mkarrmann replied:
I'm sure you're right! I looked through the history of this file for clues on how this function is actually supposed to work, but it's had the same basic structure since this file was created back in 2019 (clearly copy-and-pasted from some other file, but I stopped digging at that point).

pohly replied:
Just to be clear (because the threading is a bit odd): https://github.com/kubernetes/kubernetes/pull/124205#discussion_r1568311204 is about "let's keep WaitForAlmostAllPodsReady exactly as it was before".

mkarrmann commented on test/e2e/framework/pod/wait.go:
```diff
@@ -211,13 +218,83 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
+// that it requests the list of pods on every iteration. This is useful, for
+// example, in cluster startup, because the number of pods increases while
+// waiting. All pods that are in SUCCESS state are not counted.
+func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns string, minPods int, timeout time.Duration) error {
```
This is a new function that is substantially cleaned up and simplified. Most notably, it doesn't even include `allowedNotReadyPods`, as I think that parameter was the source of most confusion and was only used by one caller (the call in `e2e.go`, which is now using the new function above).

mkarrmann commented on test/e2e/windows/host_process.go:
```diff
@@ -657,7 +657,8 @@ var _ = sigDescribe(feature.WindowsHostProcessContainers, "[MinimumKubeletVersio
 		timeout := 3 * time.Minute
-		e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, 0, timeout)
+		err = e2epod.WaitForPodsRunningReady(ctx, f.ClientSet, f.Namespace.Name, 1, timeout)
+		framework.ExpectNoError(err)
```
One of the CI checks made me add these checks for errors

pohly replied:
And for a good reason! The test was clearly broken without this error check.

pohly replied:
Side note: in some cases I found tests that were no longer passing when adding missing error checking.

mkarrmann replied:
Agreed!

For the record, the fact that "failing to properly detect an error in this function constitutes the test that calls it being broken" is the reason that I thought it was important to fix the logic at the end of `WaitForAlmostAllPodsReady`. The logic for when the function returns an error is clearly busted, which is at least as bad as failing to check for errors.

mkarrmann commented:
Thank you @aojea and @pohly 

As suggested, I split up the function `WaitForPodsRunningReady` into two functions, one of which is very close to the original and is only called when setting up the cluster in `e2e.go` and the other is substantially simplified and called everywhere else.

I update the PR description and added more comments explain the details of my decisions.

pohly commented on test/e2e/framework/pod/wait.go:
```diff
@@ -211,13 +218,83 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
+	nOk := 0
+	badPods := []v1.Pod{}
+	otherPods := []v1.Pod{}
+	succeededPods := []string{}
```
These variables can be in the function body below, there's no need to define them outside.

pohly commented on test/e2e/framework/pod/wait.go:
```diff
@@ -211,13 +218,83 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
 
+// WaitForPodsRunningReady waits up to timeout for the following conditions:
+//  1. At least minPods Pods in Namespace ns are Running and Ready
+//  2. No Pods in Namespace ns are not either Ready, Succeeded, or Failed with a Controller
```
This second criteria is hard to understand (double negation, what is "Failed with a Controller"). How about this?

```suggestion
//  2. Only Pods owned by a controller are allowed to be in the Failed phase.
// 3. Pods in the Succeeded phase are ignored.
// 4. No Pod is in some other phase.
```

mkarrmann replied:
I definitely agree it's hard to understand. I was originally worried about another Pod Phase being added and making this comment absolete, but that's far too unlikely to be worth the convoluted English.

I changed it to `No Pods in Namespace ns are Failed and not owned by a controller or Pending`, since Pending's also an important case. Let me know if you'd like me to update it again.

mkarrmann committed:
Small clean and comment rewording of WaitForPodsRunningReady

mkarrmann commented:
Thank you for the review @pohly. I addressed all of your comments. I've kept the change I made at the end of `WaitForAlmostAllPodsReady` for, but like I said I'm more than happy to remove if you'd like.

mkarrmann committed:
Revert check at end of WaitForAlmostAllPodsReady to only consider Pending pods

mkarrmann commented:
@pohly Thanks again for the comments, I reverted my change at the end of `WaitForAlmostAllPodsReady`, so that the function now behaves exactly as `WaitForPodsRunningReady`. I kept the couple of cosmetic refactors (no change in behavior) that I had made along the way, which I assume is fine.

mkarrmann commented:
/retest

pohly reviewed (approved):
/lgtm
/approve
